File: 0202_happy_number/happy_number.py
def is_happy(n: int) -> bool:
    """
    Time: O(n)
    Space: O(n)
    """
    seen = set()
    while True:
        if n == 1:
            return True
        nums = sum([int(x)**2 for x in str(n)])
        if nums in seen:
            return False
        seen.add(nums)
        n = nums

File: 0202_happy_number/happy_number.js
/**
 * Time: O(n)
 * Space: O(n)
 */
const isHappy = n => {
  const s = new Set()
  while (true) {
    if (n === 1) return true
    let sum = n
      .toString()
      .split("")
      .reduce((a, b) => {
        a += Math.pow(+b, 2)
        return a
      }, 0)
    if (s.has(sum)) return false
    s.add(sum)
    n = sum
  }
}
